reference/reference.go

```diff
@@ -0,0 +1,181 @@
+package reference
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/digest"
+	distreference "github.com/docker/distribution/reference"
+	"github.com/docker/docker/image/v1"
+)
+
+const (
+	// DefaultTag defines the default tag used when performing images related actions and no tag string is specified
+	DefaultTag = "latest"
+	// DefaultHostname is the default built-in hostname
+	DefaultHostname = "docker.io"
+	// LegacyDefaultHostname is automatically converted to DefaultHostname
+	LegacyDefaultHostname = "index.docker.io"
+	// DefaultRepoPrefix is the prefix used for defautl repositories in default host
+	DefaultRepoPrefix = "library/"
+)
+
+// Reference is an opaque object reference identifier that may include
+// modifiers such as a hostname, name, tag, and digest.
+type Reference interface {
+	// String returns the full reference
+	String() string
+}
+
+// Named is an object with a full name
+type Named interface {
+	Reference
+	Name() string
+}
```

Comment:
Could we add a another type such as `NormalizedNamed` which can distinguish names which have been normalized. This would allow us to keep with the guarantees of the package of distinguishing references using type safety. Then this `Named` type could add a function `Normalize() (NormalizedNamed, error)`.